### PR TITLE
[Internal] Direct Package Upgrade: Refactors Code to Bump Up `Cosmos.Direct` Package to `3.31.2`

### LINKED ISSUE #3700: CosmosClient::Dispose throws SynchronizationLockException
**Describe the bug**
`CosmosClient::Dispose()` throws `SynchronizationLockException`

**To Reproduce**
We don't know exactly how to reproduce, but it happens on ASP.NET app shutdown. We have an `IDisposable` Cosmos Client registered to the ASP.NET DI that when disposed by the ASP.NET host, disposes all the different clients it created in a loop:
```cs
foreach (var client in _clients)
{
    client.Value.Value.Dispose();
}
```

Note that `.Value` is `Lazy<T>` so I suspect it could be that some of the clients above are actually created (new `CosmosClient()`) during that loop .

**Expected behavior**
`Dispose()` should not throw exceptions.

**Actual behavior**
`CosmosClient::Dispose()` sometime throws as explained above (see full screenshot below)

**Environment summary**
SDK Version: _Microsoft.Azure.Cosmos_ **3.32.0** (latest at the time of writing)
OS Version: linux (`mcr.microsoft.com/dotnet/aspnet:7.0.2-cbl-mariner2.0-distroless`)

**Additional context**
```
Unhandled exception. System.Threading.SynchronizationLockException: The lock is being disposed while still being used. 
It either is being held by a thread and/or has active waiters waiting to acquire the lock.
   at System.Threading.ReaderWriterLockSlim.Dispose(Boolean disposing)
   at Microsoft.Azure.Documents.Rntbd.LoadBalancingPartition.Dispose()
   at Microsoft.Azure.Documents.Rntbd.LoadBalancingChannel.System.IDisposable.Dispose()
   at Microsoft.Azure.Documents.Rntbd.LoadBalancingChannel.Close()
   at Microsoft.Azure.Documents.Rntbd.ChannelDictionary.Dispose()
   at Microsoft.Azure.Documents.Rntbd.TransportClient.Dispose()
   at Microsoft.Azure.Documents.StoreClientFactory.Dispose()
   at Microsoft.Azure.Cosmos.DocumentClient.Dispose()
   at Microsoft.Azure.Cosmos.ClientContextCore.Dispose(Boolean disposing)
   at Microsoft.Azure.Cosmos.ClientContextCore.Dispose()
   at Microsoft.Azure.Cosmos.CosmosClient.Dispose(Boolean disposing)
   at Microsoft.Azure.Cosmos.CosmosClient.Dispose()
   at Microsoft.Azure.XXX.Documents.Cosmos.CosmosClientFactory.DisposeAsync() in 
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.DisposeAsync()
--- End of stack trace from previous location ---
   at Microsoft.Extensions.Hosting.Internal.Host.<DisposeAsync>g__DisposeAsync|16_0(Object o)
   at Microsoft.Extensions.Hosting.Internal.Host.DisposeAsync()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at Microsoft.Azure.XXXProgram.Main(String[] args) 
```

### PR DESCRIPTION
**Pull request by kundadebdatta**
# Pull Request Template

## Description

This PR bumps up the `Cosmos.Direct` package version to `3.31.2` which addresses the below issues:

- Fixes crashing `CosmosClient` on the `Dispose()` stage.
- Marking the `Israel Central` as a public region.
- Adds the Dedicated Gateway Shard Key Header.
- Upgrade Resiliency: Refactors Rntbd Open Connection Logic to Validate Replicas Accurately. Improves the `LoadBalancingPartition.OpenChannelAsync()` located inside the `Cosmos.Direct` package, to add a validation on the maximum number of channels that is permitted to open by the user. The changes are described below in detail.

  - Code changes to refactor open connection logic in the `LoadBalancingPartition` to add max channel validation.
  - Code changes to refactor the `ChannelDictionary` to remove the check on the existing channels. This change will help to establish Rntbd connections to backend replica nodes, during the replica validation flow.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Closing issues

To automatically close an issue: closes #3700, #3908

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -3,7 +3,7 @@
 		<ClientOfficialVersion>3.34.0</ClientOfficialVersion>
 		<ClientPreviewVersion>3.34.0</ClientPreviewVersion>
 		<ClientPreviewSuffixVersion>preview</ClientPreviewSuffixVersion>
-		<DirectVersion>3.31.1</DirectVersion>
+		<DirectVersion>3.31.2</DirectVersion>
```
Comment (ealsur) [inline Directory.Build.props:6]: Add the new region to `Regions.cs`

**Review by ealsur — CHANGES_REQUESTED** [review]

**Review by NaluTripician — DISMISSED** [review]

File: Directory.Build.props (hunk 1 above)
Comment (kundadebdatta) [reply Directory.Build.props:6]: Added.

**Review by kirankumarkolli — DISMISSED** [review]

**Comment by kirankumarkolli** [comment]
@msbot please wait for @ealsur signoff

File: Directory.Build.props (hunk 1 above)
Comment (kirankumarkolli) [reply Directory.Build.props:6]: Is there test coverage for it?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/BaselineTest/TestBaseline/LinqTranslationWithCustomSerializerBaseline.TestMemberInitializer.xml [hunk 2]
```diff
@@ -0,0 +1,48 @@
+﻿<Results>
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/BaselineTest/TestBaseline/LinqTranslationWithCustomSerializerBaseline.TestMemberInitializer.xml:1]: Undo this file, it is not needed

Comment (kundadebdatta) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/BaselineTest/TestBaseline/LinqTranslationWithCustomSerializerBaseline.TestMemberInitializer.xml]: Deleted this file.

**Review by ealsur — APPROVED** [review]

**Review by kirankumarkolli — APPROVED** [review]

**Pull request merged**